File: common/src/main/java/io/github/moulberry/moulconfig/gui/component/ButtonComponent.java
package io.github.moulberry.moulconfig.gui.component;

import io.github.moulberry.moulconfig.common.KeyboardConstants;
import io.github.moulberry.moulconfig.gui.GuiComponent;
import io.github.moulberry.moulconfig.gui.GuiImmediateContext;
import io.github.moulberry.moulconfig.gui.KeyboardEvent;
import io.github.moulberry.moulconfig.gui.MouseEvent;
import lombok.var;

public class ButtonComponent extends PanelComponent {
    final Runnable onClick;

    public ButtonComponent(GuiComponent element, int insets, Runnable onClick) {
        super(element, insets);
        this.onClick = onClick;
    }

    @Override
    public void mouseEvent(MouseEvent mouseEvent, GuiImmediateContext context) {
        if (context.isHovered() && mouseEvent instanceof MouseEvent.Click) {
            var click = ((MouseEvent.Click) mouseEvent);
            if (click.getMouseState() && click.getMouseButton() == 0)
                onClick.run();
        }
    }

    @Override
    public void keyboardEvent(KeyboardEvent event, GuiImmediateContext context) {
        if (isFocused() && event.getPress() && event.getKey() == KeyboardConstants.KEY_RETURN) {
            onClick.run();
        } else
            super.keyboardEvent(event, context);
    }
}

File: common/src/main/java/io/github/moulberry/moulconfig/gui/component/ButtonComponent.kt
package io.github.moulberry.moulconfig.gui.component

import io.github.moulberry.moulconfig.common.KeyboardConstants
import io.github.moulberry.moulconfig.gui.GuiComponent
import io.github.moulberry.moulconfig.gui.GuiImmediateContext
import io.github.moulberry.moulconfig.gui.KeyboardEvent
import io.github.moulberry.moulconfig.gui.MouseEvent
import io.github.moulberry.moulconfig.gui.MouseEvent.Click

class ButtonComponent(element: GuiComponent, insets: Int, val onClick: Runnable) : PanelComponent(element, insets) {
    override fun mouseEvent(mouseEvent: MouseEvent, context: GuiImmediateContext): Boolean {
        if (context.isHovered && mouseEvent is Click) {
            val (mouseButton, mouseState) = mouseEvent
            if (mouseState && mouseButton == 0) {
                onClick.run()
                return true
            }
        }
        return false
    }

    override fun keyboardEvent(event: KeyboardEvent, context: GuiImmediateContext): Boolean {
        if (isFocused && event is KeyboardEvent.KeyPressed &&
            event.pressed && event.keycode == KeyboardConstants.enter
        ) {
            onClick.run()
            return true
        }
        return false
    }
}
